### PR TITLE
chore: Remove unused types

### PR DESCRIPTION
**Pull request by RomainMuller**
Added a linter rule to prohibit declaration of unused interfaces, data
types and enums. Unused means it is no reachable from any class. Certain
packages can declare interfaces for standard APIs, but in the general
case, an un-reachable interface / data type / enum is not useful as
it cannot be used to interact with the declared constructs.

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-servicediscovery/lib/namespace.ts [hunk 1]
```diff
@@ -39,28 +39,6 @@ export interface BaseNamespaceProps {
   readonly description?: string;
 }
 
-export interface NamespaceAttributes {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-servicediscovery/lib/namespace.ts]: Why are you deleting this? Better to add the importer then if it was missing, no?

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 2]
```diff
@@ -78,7 +78,7 @@ export interface ThrottlingPerMethod {
 /**
  * Type of Usage Plan Key. Currently the only supported type is 'API_KEY'
  */
-export enum UsagePlanKeyType {
+const enum UsagePlanKeyType {
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Any enum that has literally only ONE possible value should not be part of the exposed API anyway... It's going to be hidden (single value possible == default value), and adds no value to the API. Adding it back once a second option is added is a non-breaking chance, ergo I'm hiding this.

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 3]
```diff
@@ -197,17 +197,6 @@ export interface Dimension {
   readonly value: any;
 }
 
-/**
- * Statistic to use over the aggregation period
- */
-export enum Statistic {
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts]: Oh okay. I didn't know :)

File: packages/@aws-cdk/aws-ecr/lib/repository.ts [hunk 4]
```diff
@@ -479,3 +479,18 @@ function renderLifecycleRule(rule: LifecycleRule) {
     }
   };
 }
+
+/**
+ * Select images based on counts
+ */
+const enum CountType {
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/aws-ecr/lib/repository.ts]: Not exported = not in JSII. Then the "const enum" makes the generated code include one fewer enums, which is a "heavy" construct. I don't feel *super* strongly about it however, but it makes code a tiny bit smaller & avoids carrying a lookup map that nobody will ever use...

File: packages/@aws-cdk/aws-servicediscovery/lib/namespace.ts (hunk 1 above)
Comment (RomainMuller) [reply packages/@aws-cdk/aws-servicediscovery/lib/namespace.ts]: I did not want to ponder whether importing a namespace is actually useful or not. And if the importer is just `fromArn`, then the attributes are still useless.

**Comment by eladb** [comment]
@RomainMuller is this something you want to try and merge tomorrow AM?

**Comment by RomainMuller** [comment]
@eladb  - yeah, it's the number one item on my list today 👍🏻

File: packages/@aws-cdk/aws-stepfunctions/lib/states/state.ts [hunk 5]
```diff
@@ -498,6 +498,8 @@ function isNextable(x: any): x is INextable {
 
 /**
  * State types
+ *
+ * @internal
```
Comment (eladb) [inline packages/@aws-cdk/aws-stepfunctions/lib/states/state.ts]: If it's internal then why export? Maybe put in a different file?

File: packages/@aws-cdk/aws-config/lib/rule.ts [hunk 6]
```diff
@@ -104,6 +81,19 @@ abstract class RuleBase extends Resource implements IRule {
  * A new managed or custom rule.
  */
 abstract class RuleNew extends RuleBase {
+  /**
+   * Imports an existing rule.
+   *
+   * @param configRuleName the name of the rule
+   */
+  public static fromConfigRuleName(scope: Construct, id: string, configRuleName: string): IRule {
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/aws-config/lib/rule.ts]: Oh - merge artifact...

**Pull request merged**